addon/-private/compatibility.js

```diff
@@ -1,12 +1,67 @@
-export const wait = (function() {
-  const hasLegacyEmberTestHelpers = window.require.has('ember-test-helpers');
-  const hasLatestEmberTestHelpers = window.require.has('@ember/test-helpers');
+//
+// This is a wrapper around `@ember/test-helpers` that we need for compatibility
+// reasons. Apps and addons aren't supposed to depend directly on
+// `@ember/test-helpers`, but just use the one that their version of
+// `ember-qunit` or `ember-mocha` provides. This compatibility module does three
+// jobs for us:
+//
+// 1. Helps us determine if we are running an RFC232/268 test or not
+// 2. Provides the test helpers needed to run RFC232/268 tests
+// 3. Provides a `wait` implementation for non-RFC232/268 (legacy) tests
+//
+// To accomplish (1) and (2) we need to determine if `@ember/test-helpers` is
+// present. If it isn't, we can't possibly be running RFC232/268 tests because
+// they rely on it. If it is, then we need its `getContext()` method to see if
+// any of the the RFC232/268 setup methods have been called. So, to keep this
+// complexity encapsulated in this file, if `@ember/test-helpers` is not
+// present, we export a stub `getContext()` function that returns null,
+// indicating that we are not running RFC232/268 tests, and then the rest of the
+// addon code won't try to access any of the other `@ember/test-helpers`
+// helpers.
+//
+// To accomplish (3), we need to determine if `ember-test-helpers` is present.
+// Because it's built with legacy support, anytime `@ember/test-helpers` is
+// present, `ember-test-helpers` will also be present. So we can check for
+// `ember-test-helpers/wait` and export it if present. If it's not present, we
+// don't want to throw an exception immediately because acceptance tests don't
+// need it, so we export a `wait` function that throws an exception if and when
+// it's called.
+//
+// Once we drop support for pre-RFC268 tests, including all calls to `wait`, we
+// can delete this file and import `@ember/test-helpers` directly.
+//
 
-  if (!hasLegacyEmberTestHelpers && !hasLatestEmberTestHelpers) {
-    return function() {
-      throw new Error('ember-test-helpers or @ember/test-helpers must be installed');
+// When a module imports `require`, it gets a dynamically generated module that
+// handles relative imports correctly, so there's no way to get at it to stub it
+// from another module/test. So instead we use the global require, which is only
+// available via window.require, so our tests can stub it out.
+const { require } = window;
+
+let helpers;
+
+if (require.has('@ember/test-helpers')) {
+  helpers = require('@ember/test-helpers');
+} else {
+  helpers = {
+    getContext() {
+      return null;
     }
-  }
+  };
+}
+
+if (require.has('ember-test-helpers/wait')) {
+  helpers.wait = require('ember-test-helpers/wait').default;
```

Comment:
@simonihmig this is to support legacy apps. `ember-test-helpers`'s `wait` hasn't always existed. `ember-cli-page-object` previously had `ember-test-helpers` as a runtime dependency to be sure `wait` was present. Now it doesn't, so apps with an old enough `ember-qunit`/`ember-test-helpers` won't have the `wait` helper present and will run into that exception you hit instead of the nice descriptive error message that you didn't get :(

I'm confused about exactly what you're seeing. Are you running `ember-cli-page-object`'s unit tests? Or your own tests using this branch? What version of `@ember/test-helpers`/`ember-test-helpers` do you have installed? `loader.js` isn't that complicated -- if you want to peel back an additional layer you could step into the require call and maybe see why it's giving you back an empty object...

Reply:
I am running my apps tests, using the master branch of `ember-mocha`. Should be the latest version of `@ember/test-helpers`, but will double check. 

> if you want to peel back an additional layer you could step into the require call and maybe see why it's giving you back an empty object...

Ok, will do that, but not before tomorrow...

Reply:
Okay. I haven't seen anything like that, and the `ember-cli-page-object` tests run all the relevant `ember-try` scenarios, so I'm not sure how to investigate with more info or a sample project that reproduces it. So let me know what you find!

Reply:
@bendemboski so here is what I got...

First of all I have `@ember/test-helpers` 0.7.18 installed...

When I step into the require call, I can see that the module is in a `pending` state, and the exports are empty:

![image](https://user-images.githubusercontent.com/1325249/36599413-8fa5b950-18af-11e8-914c-7294887567f1.png)

As stated before, when I explicitly import `ember-test-helpers/wait` in that file, it works, and as you can see the module is in the `finalized` state and the exports are all there:

![image](https://user-images.githubusercontent.com/1325249/36600503-c0543d62-18b2-11e8-843a-b9886a21fa79.png)

This is happening only in a few tests, so I *suppose*  this is depending on the order the modules are required: when `wait` has been required before by some other module, it is already in the `finalized` state, and the require gets the correct exports. But when the `require` call in question here is the first to ever need the module, and it has not declared `ember-test-helpers/wait` as a dependency (by importing it), it will see a `pending` module and the exports are still empty.

Reply:
I'll try to get a repro of this when I have the chance (probably later today or tomorrow), but if you have time, could you try changing this line to

```javascript
helpers.wait = (...args) => require('ember-test-helpers/wait').default(...args);
```

and see if it works for you? I'm hoping that moving the require to "runtime" rather than "module load time" (if you know what I mean) will fix it.

Reply:
Just tried it quickly, and yes, it indeed fixes the problem! 😀

I probably miss something important, but what would be the problem if `e-c-p-o` would simply add the latest `@ember/test-helpers` as a dependency? (so you could simply import it and not require all of this)

Reply:
It used to. @magistrula observed that (maybe due to an `ember-cli` bug or maybe an intended feature?) if the app also specified a version of `@ember/test-helpers`, the addon's version of `@ember/test-helpers` was overriding the version the app specified (see #361 and #366). This seems like risky behavior to accept in order to patch up a very uncommon case (an app having a version of `ember-test-helpers` in the pretty old and narrow window when it didn't have `wait`).

Also, when I asked in slack, I think it was @Turbo87 and/or @rwjblue that told me that apps and addons can and should just depend on `ember-qunit` or `ember-mocha` to supply `@ember/test-helpers`, and not list it as a dependency themselves.

Anyway, thanks a bunch for digging in -- I'll make that change to defer the require execution until after modules are loaded.